### DISCUSSION TITLE
InfluxDB reporting on commands *RESOLVED*

### DISCUSSION THREAD

**Discussion by pkkrusty** [arendst/Tasmota#15257] [Ideas]
I've been using and enjoying InfluxDB integration, for visualizing time-based numerical data and seeing trends of power state of relays. However, I'd like to track the numerical JSON values of recurring commands, in my instance `ping`

The mqtt response I get from my once per minute ping is `MQT: router-manager/tele/RESULT = {"Ping":{"8.8.8.8":{"Reachable":true,"IP":"8.8.8.8","Success":4,"Timeout":0,"MinTime":48,"MaxTime":57,"AvgTime":53}}}`

I'd like to be able to track the 2nd-level JSON response of Reachable and/or Success so I can see when and for how long my internet connection dies. 

Am I missing something in my config, such that the InfluxDB driver can already do that? When I pull up available metrics for this particular ESP8266 chip, I get the standard STATE ones like Uptime, Heap, Sleep and the power state ones like Power1, Power2, etc. Looking at the driver code, I think that's all that's available.

Not sure what it would take to integrate reporting of Command RESULT statements into the driver, but seems like that would extend functionality for tracking things like triggers, and in my case, WAN status.

I could build something in Node-Red, but the InfluxDB driver is a lot cleaner.

**Comment by pkkrusty** [comment]
Or maybe I can have it trigger `on ping#8.8.8.8#SUCCESS==0 do publish router-manager/tele/STATE {"Time":%timestamp%,"reachable":%value%}`

Would the influxdb driver pick up on that `tele` publish?

**Comment by barbudor** [comment]
The driver is meant to automatically publish fields from the STATE and SENSOR messages to InfluxDB. It does not support the results from commands. If this was supported, I think it would generate a massive number of records mostly useless.
The driver has it's own polling interval and request from the other drivers the JSON string for STATE and SENSOR, then it parses it and publish to InfluxDB.
With regards to your 2nd message, the driver won't neither be able to capture something from the Publish command.


However you don't need the driver to publish to InfluxDB by using the `WebQuery` command and InfluxDB Line Protocol
The general syntax would be :
`WebQuery http://your.server.ip:8086/write?db=databasename POST measurementname,tagname=tagvalue field1name=field1value,field2name=field2value`

Beware of the spaces in the body they are critical
Details of Line Protocol : https://docs.influxdata.com/influxdb/v1.8/write_protocols/line_protocol_tutorial/

**Reply by pkkrusty** [reply]
Ah super, will give that a try. thanks.

EDIT: Works great!

**Reply by barbudor** [reply]
Beware with the timestamp. The Line Protocol optionnaly supports a timestamp but if not specified it will use the clock from the server it self to timestamp the data. In general that means very little difference
But If you absolutely want to use Tasmota's timestamp which is in seconds, you would need to add `000` because InfluxDB requires a timestamp in milliseconds.
Also it unfortunately doesn't support a timestamp in ISO format

**Reply by pkkrusty** [reply]
Yeah, I'm letting it use the system timestamp, since I have a local NTP server that everything syncs to with microsecond accuracy. (GPS with PPS signal)

**Comment by pkkrusty** [comment]
Noticed a commit by @arendst this evening regarding non-teleperiod Influx reporting. Is there anywhere in docs or discussion that gives more details for what this entails? The commit is only 1 hour old, so I'm probably jumping the gun, but I'm not good enough at reading the code to see what's going on.

**Reply by arendst** [reply]
Another commit regarding the same is introducing command ``IfxSensor 0/1`` to enable this feature.

In addition to the "normal" teleperiod data which is send to influxdb this command adds support for sensor data which is provided by drivers directly so not requested during teleperiod time. Think about a weight change on a HX711 scale.

As it cannot distinguish between valid influx data you might see influx error messages like 

```
22:19:20.451-024 MQT: stat/hema1/STATUS3 = {"StatusLOG":{"SerialLog":3,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"loghost","LogPort":514,"SSId":["indebuurt_IoT",""],"TelePeriod":10,"Resolution":"558186C0","SetOption":["00008009","2805C80001000600003C5A0A000000000000","00000080","00006000","00014000"]}}
22:19:20.496-022 IFX: Error {"error":"partial write: unable to parse 'setoption2,device=hema1,sensor=statuslog value=2805C80001000600003C5A0A000000000000': invalid number dropped=0"}
```

Filtering out these errors just needs way too much code so use the new command with care.

**Reply by barbudor** [reply]
@arendst
Do I get it right that with `IfxSensor 1` every message that goes through the RuleProcess will be published to InfluxDB regardless of what it contains ? Like logging sensors at max speed but also each and every command result ?

**Reply by arendst** [reply]
Yes and No.

When enabled it reports every call to ``XdrvRulesProcess(0)``. This is used by ad-hoc rule processing and, on request from users, SOME commands like ``status``. So indeed you could get false reports but once a system is configured most false reports are not being generated.

As a test enable logging mode 4 and see when and what influxdb publish. You might need to execute command ``IfxLog 4`` too.

**Reply by arendst** [reply]
Oh bugger, it reports on all commands indeed. 

As long as no commands are given it shouldn't be a problem ;-)

**Reply by arendst** [reply]
I think I better change it to only report for ``MqttPublishTeleSensor``.

**Reply by arendst** [reply]
Will change now. Hold on.

**Reply by arendst** [reply]
Done.

https://github.com/arendst/Tasmota/commit/721bb4143bd5b36fa453beace57c301e40daa7a8

**Reply by barbudor** [reply]
I was thinking it could be interresting to have an `IfxPublish` (or IfxPost) command as a wrapper to the WebQuery above so that
`WebQuery http://your.server.ip:8086/write?db=databasename POST measurementname,tagname=tagvalue field1name=field1value,field2name=field2value`
could be replaced by the more handy
`IfxPublish measurementname,tagname=tagvalue field1name=field1value,field2name=field2value`
The command could also automatically add the millisecond epoch timestamp to have the record timestamped by the ESP and not the receiving server.
I can do such PR if you think it would be interresting

**Reply by arendst** [reply]
Yes. Pls provide PR.

**Reply by pkkrusty** [reply]
> I was thinking it could be interresting to have an `IfxPublish` (or IfxPost) command as a wrapper to the WebQuery above so that `WebQuery http://your.server.ip:8086/write?db=databasename POST measurementname,tagname=tagvalue field1name=field1value,field2name=field2value` could be replaced by the more handy `IfxPublish measurementname,tagname=tagvalue field1name=field1value,field2name=field2value` The command could also automatically add the millisecond epoch timestamp to have the record timestamped by the ESP and not the receiving server. I can do such PR if you think it would be interresting

@barbudor Could we add a variable to define the retention policy per device or have two or three choices per device? Could do it with the webquery, but for my weather data, I'd like infinite (or maybe 5 years) retention policy, but for power/network monitoring, 30d is sufficient. Battery charging, one year, etc. Most devices are probably only doing one thing, so IfxRetention (0 is default, otherwise integer=number of days retention) would be sufficient for most people.

*edit Although maybe the passed variable in the webquery needs to match defined retention policies...in which case the variable would have to be a string...like in the documentation.

`http://your.server.ip:8086/write?db=databasename&rp=autogen POST measurementname,tagname=tagvalue field1name=field1value,field2name=field2value`

**Reply by barbudor** [reply]
@pkkrusty
The duration of the retention policy itself should be configured on the InfluxDB server
The only element that you can pass to the write command is the name of a predefined retention policy (if non specified, InfluxDB use the one tagged default)

I will have a look at how this could be made configurable on the command
I've still not worked on the PR, probably I will find time this week-end
The idea is to have an `IfxWrite` command that would just wrap the WebQuery command up to the POST method. The rest of the line would remains the standard line-protocol syntax

**Reply by pkkrusty** [reply]
Yeah, so would need to have that second part of the web query defined in the wrapper (&rp=definedpolicy)

**Reply by pkkrusty** [reply]
Submitted #15513 to add retention policy option to the driver. Not sure how that would integrate with your work on the ifxwrite command.

**Reply by barbudor** [reply]
Sorry haven't progress yet on that one. I'll restart on it once your PR has been merge
